File: django_project/frontend/src/models/Property.ts
import ParcelInterface from "./Parcel"

export default interface PropertyInterface {
    id: number,
    name: string,
    owner: string,
    owner_email?: string,
    property_type: string,
    property_type_id?: number,
    open: string,
    open_id?: number,
    province: string,
    province_id?: number,
    size?: number,
    organisation?: string,
    organisation_id?: number,
    parcels?: ParcelInterface[],
    bbox?: number[],
    centroid?: number[],
    short_code?: string,
    boundary_search_session?: string,
    boundary_source?: string
}

export const createNewProperty = ():PropertyInterface => {
    let _result:PropertyInterface = {
        id: 0,
        name: '',
        owner: '',
        owner_email: '',
        property_type: '',
        province: '',
        size: 0,
        organisation: '',
        open: '',
        parcels: [],
        property_type_id: 0,
        province_id: 0,
        organisation_id: 0,
        open_id: 0,
    }
    return _result
}

export interface PropertyValidation {
    name?: boolean,
    email?: boolean,
    property_type?: boolean,
    province?: boolean,
    organisation?: boolean
}

export interface PropertyTypeInterface {
    id: number,
    name: string
}

export interface ProvinceInterface {
    id: number,
    name: string
}

export interface OpenCloseInterface {
    id: number,
    name: string
}

export const SELECT_SOURCE_TYPE = 'selection'
export const DIGITISE_SOURCE_TYPE = 'digitise'
export const BOUNDARY_FILE_SOURCE_TYPE = 'boundary_file'

export interface BoundarySearchResultInterface {
    session: string,
    status: string,
    progress: number,
    parcels?: ParcelInterface[],
    province: string,
    property_size_ha: number,
    type: string,
    bbox: number[],
    upload_file_names?: string[]
}

File: django_project/frontend/serializers/property.py
"""Serializer for property classes."""
from rest_framework import serializers
from frontend.serializers.common import NameObjectBaseSerializer
from property.models import (
    PropertyType,
    Province,
    Property,
    Parcel
)


class PropertyTypeSerializer(NameObjectBaseSerializer):
    """Property Type Serializer."""

    class Meta:
        model = PropertyType
        fields = NameObjectBaseSerializer.Meta.fields


class PropertyTypeColourSerializer(serializers.ModelSerializer):
    """Property Type Serializer with Colour."""

    class Meta:
        model = PropertyType
        fields = [
            'id',
            'name',
            'colour'
        ]


class ProvinceSerializer(NameObjectBaseSerializer):
    """Province Serializer."""

    class Meta:
        model = Province
        fields = NameObjectBaseSerializer.Meta.fields


class PropertySerializer(serializers.ModelSerializer):
    """Property Serializer."""

    owner = serializers.SerializerMethodField()
    owner_email = serializers.SerializerMethodField()
    property_type = serializers.SerializerMethodField()
    open = serializers.SerializerMethodField()
    province = serializers.SerializerMethodField()
    organisation = serializers.SerializerMethodField()
    size = serializers.SerializerMethodField()

    def get_owner(self, obj: Property):
        name = (
            f'{obj.created_by.first_name} {obj.created_by.last_name}'.strip()
        )
        return name if name else obj.created_by.username

    def get_owner_email(self, obj: Property):
        return obj.created_by.email

    def get_property_type(self, obj: Property):
        return obj.property_type.name

    def get_open(self, obj: Property):
        return obj.open.name if obj.open else ''

    def get_province(self, obj: Property):
        return obj.province.name

    def get_organisation(self, obj: Property):
        return obj.organisation.name

    def get_size(self, obj: Property):
        return obj.property_size_ha


    class Meta:
        model = Property
        fields = [
            'id', 'name', 'owner', 'owner_email',
            'property_type', 'property_type_id',
            'province', 'province_id',
            'open', 'open_id',
            'size', 'organisation', 'organisation_id',
            'short_code', 'boundary_source'
        ]


class ParcelSerializer(serializers.ModelSerializer):
    """Parcel Serializer."""
    layer = serializers.SerializerMethodField()
    type = serializers.SerializerMethodField()
    cname = serializers.SerializerMethodField()
    id = serializers.SerializerMethodField()

    def get_id(self, obj: Parcel):
        return obj.source_id

    def get_cname(self, obj: Parcel):
        return obj.sg_number

    def get_layer(self, obj: Parcel):
        return obj.source

    def get_type(self, obj: Parcel):
        return obj.parcel_type.name.lower()

    class Meta:
        model = Parcel
        fields = [
            'id', 'layer', 'type', 'cname'
        ]


class PropertyDetailSerializer(PropertySerializer):
    """Property with more details."""
    parcels = serializers.SerializerMethodField()
    bbox = serializers.SerializerMethodField()
    centroid = serializers.SerializerMethodField()

    def get_parcels(self, obj: Property):
        parcels = obj.parcel_set.all()
        return ParcelSerializer(parcels, many=True).data

    def get_bbox(self, obj: Property):
        if obj.geometry is None:
            return []
        return list(obj.geometry.envelope.extent)

    def get_centroid(self, obj: Property):
        if obj.geometry is None:
            return []
        return list(obj.geometry.centroid)

    class Meta:
        model = Property
        fields = [
            'id', 'name', 'owner', 'owner_email',
            'property_type', 'property_type_id',
            'open', 'open_id',
            'province', 'province_id',
            'size', 'organisation', 'organisation_id',
            'parcels', 'bbox', 'centroid', 'short_code',
            'boundary_source'
        ]


class PropertySearchSerializer(PropertyDetailSerializer):
    """Return id, name, bbox of property."""
    id = serializers.SerializerMethodField()
    type = serializers.SerializerMethodField()
    name = serializers.SerializerMethodField()

    def get_id(self, obj: Property):
        return f'property-{obj.id}'

    def get_type(self, obj: Property):
        return 'property'

    def get_name(self, obj: Property):
        return f'{obj.name} ({obj.short_code})'

    class Meta:
        model = Property
        fields = [
            'id', 'type', 'name', 'bbox'
        ]
